pkg/image/registry/imagestreamimport/rest.go

```diff
@@ -207,6 +207,18 @@ func (r *REST) Create(ctx apirequest.Context, obj runtime.Object, createValidati
 		return nil, kapierrors.NewInternalError(err)
 	}
 
+	// check imported images status. If we get authentication error (401), try import same image without authentication.
+	// Docker registry gives 401 on public images if you have wrong secret in your secret list.
+	for _, image := range isi.Status.Images {
+		if image.Status.Code == http.StatusUnauthorized {
+			importCtx := importer.NewContext(r.transport, r.insecureTransport).WithCredentials(nil)
+			imports := r.importFn(importCtx)
+			if err := imports.Import(ctx.(gocontext.Context), isi, stream); err != nil {
```

Comment:
one problem I see with this is if the image really is private, does this turn what was a legitimate 401 into a 404 error when you make the import call with no credentials?  (because docker returns 404 for private registries?)

That is, if the scenario is:  i'm trying to import from a private registry and I provide invalid credentials, previously I would get a proper 401 and know my credentials were (probably) wrong.  Is that what will still happen with this change, or will I now see a 404 error and be confused?

Reply:
I did the bunch of tests and checking how this "awesome" `docker registry API` behaves and when we do this (spent way too much time on this...). This fix for me looked least intrusive and did the job. 

I think your case is covered because it still gives 401 but it gives different error message so in `importer.go` it gets assigned different error message.  

There is chance it might behave diferently if it will be used with non `docker distribution` based docker registry API's (Nexus in example) which will give different error messages. But in this case all our error handling logic is flawed. 

The current behavior of our IS importer with private and public images with wrong secret:
```
-bash-4.4#  oc create secret docker-registry dummy-secret1 --docker-server=docker.io --docker-username=dummy1 --docker-password=dummy1 --docker-email==dummy1@example.com
secret "dummy-secret1" created    
##### PUBLIC IMAGE  
-bash-4.4# oc import-image example --from=openshift/hello-openshift --confirm

  ! error: Import failed (InternalError): Internal error occurred: Get https://registry-1.docker.io/v2/openshift/hello-openshift/manifests/latest: unauthorized: incorrect username or password
      Less than a second ago

error: tag latest failed: Internal error occurred: Get https://registry-1.docker.io/v2/openshift/hello-openshift/manifests/latest: unauthorized: incorrect username or password

##### PRIVATE IMAGE  
-bash-4.4# oc import-image example1 --from=mangirdas/kafka-producer:latest --confirm
The import completed with errors.

  ! error: Import failed (InternalError): Internal error occurred: Get https://registry-1.docker.io/v2/mangirdas/kafka-producer/manifests/latest: unauthorized: incorrect username or password
      Less than a second ago

error: tag latest failed: Internal error occurred: Get https://registry-1.docker.io/v2/mangirdas/kafka-producer/manifests/latest: unauthorized: incorrect username or password
```

With fix:
```
-bash-4.4#  oc create secret docker-registry dummy-secret1 --docker-server=docker.io --docker-username=dummy1 --docker-password=dummy1 --docker-email==dummy1@example.com                                          
secret "dummy-secret1" created  
##### PUBLIC IMAGE                                                                                                                                                                                     
-bash-4.4# oc import-image example --from=openshift/hello-openshift --confirm                                                                                                                                      
The import completed successfully.                                                                                                                                                                                 

latest                                              
  tagged from openshift/hello-openshift             

  * openshift/hello-openshift@sha256:1a7a40888035f95c10cde81f04c10aa5919a41766e4680452918fc5591d7b02a    
      Less than a second ago                        

##### PRIVATE IMAGE  
-bash-4.4# oc import-image example1 --from=mangirdas/kafka-producer:latest --confirm                     
The import completed with errors.                   

latest                                              
  tagged from mangirdas/kafka-producer:latest       

  ! error: Import failed (Unauthorized): you may not have access to the Docker image "mangirdas/kafka-producer:latest"                                                                                             
      Less than a second ago                        

error: tag latest failed: you may not have access to the Docker image "mangirdas/kafka-producer:latest"  
```


Without secret:
```
##### PUBLIC IMAGE  
-bash-4.4# oc import-image example --from=openshift/hello-openshift --confirm
The import completed successfully.

latest
  tagged from openshift/hello-openshift

  * openshift/hello-openshift@sha256:1a7a40888035f95c10cde81f04c10aa5919a41766e4680452918fc5591d7b02a
      Less than a second ago

```

##### PRIVATE IMAGE  
```
-bash-4.4# oc import-image example1 --from=mangirdas/kafka-producer:latest --confirm

latest
  tagged from mangirdas/kafka-producer:latest

  ! error: Import failed (Unauthorized): you may not have access to the Docker image "mangirdas/kafka-producer:latest"
      Less than a second ago

error: tag latest failed: you may not have access to the Docker image "mangirdas/kafka-producer:latest"
```

My secret is OK but the image is in the private repo of other person account:
```
oc create secret docker-registry dummy-secret1 --docker-server=docker.io --docker-username=mangirdas --docker-password=dummy1 --docker-email=mangirdas@example.com

-bash-4.4# oc import-image example2 --from=juozasa/hellonode:latest --confirm

latest
  tagged from juozasa/hellonode:latest

  ! error: Import failed (Unauthorized): you may not have access to the Docker image "juozasa/hellonode:latest"
      Less than a second ago

error: tag latest failed: you may not have access to the Docker image "juozasa/hellonode:latest"
```